utils/benchmarking/eventing/timeseriesdb/timeseriesdb_test.go

```diff
@@ -0,0 +1,147 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+package main
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	ctrdlog "github.com/containerd/containerd/log"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/ease-lab/vhive/utils/benchmarking/eventing/proto"
+)
+
+const SleepDuration = 200 * time.Millisecond
+
+func TestDefaultEndToEnd(t *testing.T) {
+	var server Server
+
+	// Parameters
+	// ================================================================
+	exDef := proto.ExperimentDefinition{CompletionEventDescriptors: []*proto.CompletionEventDescriptor{
+		{
+			AttrMatchers: map[string]string{
+				"type":   "greeting",
+				"source": "consumer",
+			},
+		},
+	}}
+	aInvokedOn := time.Now().UTC()
+	eventA1 := proto.Event{
+		VHiveMetadata: &proto.VHiveMetadata{Id: "A", InvokedOn: timestamppb.New(aInvokedOn)},
+		Attributes: map[string]string{
+			"id": "A",
+			"source": "producer",
+			"type": "greeting",
+		},
+		Data: []byte("some JSON data..."),
+	}
+	eventA2 := proto.Event{
+		VHiveMetadata: &proto.VHiveMetadata{Id: "A", InvokedOn: timestamppb.New(aInvokedOn)},
+		Attributes: map[string]string{
+			"id": "A",
+			"source": "consumer",
+			"type": "greeting",
+		},
+		Data: []byte("some JSON data..."),
+	}
+
+	// Scenario
+	// ================================================================
+	// invoker calls StartExperiment():
+	_, err := server.StartExperiment(context.Background(), &exDef)
+	if err != nil {
+		t.Error("StartExperiment", err)
+	}
+
+	// relay calls RegisterEvent() upon the CloudEvent (A1) created by producer:
+	_, err = server.RegisterEvent(context.Background(), &eventA1)
+	if err != nil {
+		t.Error("RegisterEvent(A1)", err)
+	}
+
+	// consumer works for a while:
+	time.Sleep(SleepDuration)
+
+	// relay calls RegisterEvent() upon the CloudEvent (A2) created the consumer:
+	_, err = server.RegisterEvent(context.Background(), &eventA2)
+	if err != nil {
+		t.Error("RegisterEvent(A2)", err)
+	}
+
+	// invoker calls EndExperiment() to collect the results:
+	res, err := server.EndExperiment(context.Background(), nil)
+	if err != nil {
+		t.Error("EndExperiment", err)
+	}
+
+	// Assertions
+	// ================================================================
+	if len(res.Invocations) != 1 {
```

Comment:
here and everywhere below use Golang's testify/require module instead of conditionals
https://pkg.go.dev/github.com/stretchr/testify/require#hdr-Example_Usage

it's much more concise